File: day_3/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

const width int = 31

func main() {
	input, err := ioutil.ReadFile("./input.txt")
	if err != nil {
		panic(err)
	}

	lines := strings.Split(string(input), "\n")

	fmt.Printf("Hit trees on the way down (%d right, %d down): %d \n", 1, 1, calcHits(1, 1, lines))
	fmt.Printf("Hit trees on the way down (%d right, %d down): %d \n", 3, 1, calcHits(3, 1, lines))
	fmt.Printf("Hit trees on the way down (%d right, %d down): %d \n", 5, 1, calcHits(5, 1, lines))
	fmt.Printf("Hit trees on the way down (%d right, %d down): %d \n", 7, 1, calcHits(7, 1, lines))
	fmt.Printf("Hit trees on the way down (%d right, %d down): %d \n", 1, 2, calcHits(1, 2, lines))
}

func calcHits(rightSteps int, downSteps int, lines []string) int {
	rightPos := rightSteps
	downPos := downSteps
	treesHit := 0

	for downPos < len(lines) {
		if lines[downPos][rightPos] == '#' {
			treesHit++
		}

		downPos += downSteps
		rightPos = (rightPos + rightSteps) % width
	}

	return treesHit
}

File: day_3/main.py
lines = []
with open("input.txt") as f:
    lines = f.readlines()

width = 31

next_pos = 3
hit_trees = 0
line_counter = 2

def hit_symbol(char):
    if char == "#":
        return "X"
    return "O"

for line in lines[1:]:
    if line[next_pos] == '#':
        hit_trees += 1

    next_pos = (next_pos + 3) % width
    line_counter += 1
    
    symbol = hit_symbol(line[next_pos])
    hit_line = line[:next_pos] + symbol + line[next_pos + 1:]
    print(hit_line)

print("Hit trees on the way down: " + str(hit_trees))
